lib/vm/src/export.rs

```diff
@@ -31,6 +31,16 @@ pub struct ExportFunction {
     pub address: *const VMFunctionBody,
     /// Pointer to the containing `VMContext`.
     pub vmctx: VMFunctionEnvironment,
+    /// Function pointer to `WasmerEnv::finish(&mut self, instance: &Instance)`.
+    ///
+    /// This function is called to finish setting up the environment after
+    /// we create the `api::Instance`.
+    // META: if you have a better idea of how to get this function to where it
+    // needs to be, please let me know.
+    pub function_ptr: Option<
+        // TODO: review the return value here...
+        fn(*mut std::ffi::c_void, *const std::ffi::c_void) -> Result<(), *mut std::ffi::c_void>,
+    >,
```

Comment:
I don't think `function_ptr` should live in the `ExportFunction`. Perhaps we need a better abstraction in the engine itself.
I don't think it should leak to the VM (other than when we init the instance)

Reply:
Well, the problem is that we're getting it from a `Resolver` (specifically the import object);

Following the types we get:

```Rust
    pub fn ImportObject::get_export(&self, module: &str, name: &str) -> Option<Export>
    fn NameResolver::resolve_by_name(&self, module: &str, name: &str) -> Option<Export>
```

```Rust
/// The value of an export passed from one instance to another.
#[derive(Debug, Clone)]
pub enum Export {
    /// A function export value.
    Function(ExportFunction),

    /// A table export value.
    Table(ExportTable),

    /// A memory export value.
    Memory(ExportMemory),

    /// A global export value.
    Global(ExportGlobal),
}
```

So I'm not sure where else we could put it if this is how we're getting the functions

Reply:
Yeah, perhaps we need two separate types. `Export` and `VMExport` (`VMExport` will be what now `Export` is)

Reply:
For example, `ExportFunction` now takes:
```
    /// The function kind (specifies the calling convention for the function).
    pub kind: VMFunctionKind,
    /// Address of the function call trampoline owned by the same VMContext that owns the VMFunctionBody.
    /// May be None when the function is a host-function (FunctionType == Dynamic or vmctx == nullptr).
    pub call_trampoline: Option<VMTrampoline>,
```

But those fields are really only needed for the engine